File: backend/src/steps/analyze.ts
import fs from "node:fs/promises";
import path from "node:path";
import Mustache from "mustache";
import { JobAnalysis, JobAnalysisSchema } from "../schemas/jobAnalysis.js";
import { loadResumeEnv, MissingApiKeyError, ResumeEnv } from "../util/env.js";
import { ensureDir } from "../util/fsUtil.js";
import { nowIso } from "../util/timeUtil.js";
import { extractJsonObject, perplexityChat } from "../clients/perplexity.js";

export type AnalyzeStepOpts = {
  repoRoot: string;
  jobPath: string;
  runDir: string;
  env?: ResumeEnv;
};

export type AnalyzeStepResult = {
  jobAnalysisPath: string;
  manualPromptPath?: string;
  via: "perplexity" | "manual";
};

const SYSTEM_PROMPT =
  "You are a job-description analyst. You MUST output a single JSON object matching " +
  "the schema described in the user prompt and nothing else. No prose, no markdown fences.";

async function readSearchPromptTemplate(repoRoot: string): Promise<string> {
  const promptPath = path.join(repoRoot, "templates", "search.prompt.md");
  return fs.readFile(promptPath, "utf8");
}

function renderSearchPrompt(template: string, jobText: string, jobPath: string): string {
  return Mustache.render(
    template,
    { JOB_PATH: jobPath, JOB_TEXT: jobText },
    undefined,
    { escape: (v: unknown) => String(v) }
  );
}

export async function analyzeStep(opts: AnalyzeStepOpts): Promise<AnalyzeStepResult> {
  const env = opts.env ?? (await loadResumeEnv(opts.repoRoot));
  const inputsDir = path.join(opts.runDir, "inputs");
  await ensureDir(inputsDir);

  const jobText = await fs.readFile(opts.jobPath, "utf8");
  const jobRelPath = path.relative(opts.repoRoot, opts.jobPath);
  const template = await readSearchPromptTemplate(opts.repoRoot);
  const userPrompt = renderSearchPrompt(template, jobText, jobRelPath);

  const jobAnalysisPath = path.join(inputsDir, "job_analysis.json");

  if (env.RESUME_USE_PERPLEXITY_RESEARCH) {
    if (!env.PERPLEXITY_API_KEY) {
      throw new MissingApiKeyError("RESUME_USE_PERPLEXITY_RESEARCH", "PERPLEXITY_API_KEY");
    }
    const raw = await perplexityChat({
      apiKey: env.PERPLEXITY_API_KEY,
      system: SYSTEM_PROMPT,
      user: userPrompt,
      responseFormat: "json_object"
    });
    const jsonText = extractJsonObject(raw);
    const parsed = JSON.parse(jsonText) as Partial<JobAnalysis>;
    const merged: JobAnalysis = JobAnalysisSchema.parse({
      schemaVersion: 1,
      generatedAt: nowIso(),
      jobSource: parsed.jobSource ?? jobRelPath,
      tagsForGraphQL: parsed.tagsForGraphQL ?? { domains: [], stack: [], kinds: ["experience", "education", "project"], groups: [] },
      tagsForTokenization: parsed.tagsForTokenization ?? [],
      summaryForGeneration: parsed.summaryForGeneration ?? "",
      rawProviderResponse: raw
    });
    await fs.writeFile(jobAnalysisPath, JSON.stringify(merged, null, 2) + "\n", "utf8");
    return { jobAnalysisPath, via: "perplexity" };
  }

  const manualPromptPath = path.join(inputsDir, "job_analysis.MANUAL.md");
  const manualPrompt =
    "<!--\n" +
    "Manual Step 1: paste this prompt into Perplexity (or any chat UI),\n" +
    "then save the JSON response to inputs/job_analysis.json next to this file.\n" +
    "The CLI's `retrieve` and `rank` subcommands will pick it up automatically.\n" +
    "-->\n\n" +
    userPrompt;
  await fs.writeFile(manualPromptPath, manualPrompt, "utf8");

  try {
    await fs.access(jobAnalysisPath);
  } catch {
    const stub: JobAnalysis = {
      schemaVersion: 1,
      generatedAt: nowIso(),
      jobSource: jobRelPath,
      tagsForGraphQL: {
        domains: [],
        stack: [],
        kinds: ["experience", "education", "project"],
        groups: []
      },
      tagsForTokenization: [],
      summaryForGeneration: "TODO: fill summaryForGeneration via the manual prompt."
    };
    await fs.writeFile(jobAnalysisPath, JSON.stringify(stub, null, 2) + "\n", "utf8");
  }

  return { jobAnalysisPath, manualPromptPath, via: "manual" };
}

File: backend/src/resume_orchestrator/steps/analyze.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

from jinja2 import Environment

from ..clients.perplexity import PerplexityChatOpts, extract_json_object, perplexity_chat
from ..schemas.job_analysis import JobAnalysis, TagsForGraphQL
from ..util.env import MissingApiKeyError, ResumeEnv, load_resume_env
from ..util.fs_util import ensure_dir
from ..util.time_util import now_iso

SYSTEM_PROMPT = (
    "You are a job-description analyst. You MUST output a single JSON object matching "
    "the schema described in the user prompt and nothing else. No prose, no markdown fences."
)


@dataclass(slots=True)
class AnalyzeStepResult:
    job_analysis_path: Path
    via: Literal["perplexity", "manual"]
    manual_prompt_path: Path | None = None


def _render_search_prompt(template: str, job_text: str, job_path: str) -> str:
    env = Environment(autoescape=False, keep_trailing_newline=True)
    return env.from_string(template).render(JOB_PATH=job_path, JOB_TEXT=job_text)


async def analyze_step(
    *,
    repo_root: str | Path,
    job_path: str | Path,
    run_dir: str | Path,
    env: ResumeEnv | None = None,
) -> AnalyzeStepResult:
    repo_root_p = Path(repo_root)
    job_path_p = Path(job_path)
    run_dir_p = Path(run_dir)
    env = env or load_resume_env(repo_root_p)

    inputs_dir = ensure_dir(run_dir_p / "inputs")

    job_text = job_path_p.read_text(encoding="utf-8")
    try:
        job_rel_path = job_path_p.relative_to(repo_root_p).as_posix()
    except ValueError:
        job_rel_path = job_path_p.as_posix()

    template_path = repo_root_p / "templates" / "search.prompt.md"
    template = template_path.read_text(encoding="utf-8")
    user_prompt = _render_search_prompt(template, job_text, job_rel_path)

    job_analysis_path = inputs_dir / "job_analysis.json"

    if env.RESUME_USE_PERPLEXITY_RESEARCH:
        if not env.PERPLEXITY_API_KEY:
            raise MissingApiKeyError("RESUME_USE_PERPLEXITY_RESEARCH", "PERPLEXITY_API_KEY")
        raw = await perplexity_chat(
            PerplexityChatOpts(
                api_key=env.PERPLEXITY_API_KEY,
                system=SYSTEM_PROMPT,
                user=user_prompt,
                response_format="json_object",
            )
        )
        json_text = extract_json_object(raw)
        parsed = json.loads(json_text)
        merged = JobAnalysis.model_validate(
            {
                "schemaVersion": 1,
                "generatedAt": now_iso(),
                "jobSource": parsed.get("jobSource") or job_rel_path,
                "tagsForGraphQL": parsed.get("tagsForGraphQL")
                or {
                    "domains": [],
                    "stack": [],
                    "kinds": ["experience", "education", "project"],
                    "groups": [],
                },
                "tagsForTokenization": parsed.get("tagsForTokenization") or [],
                "summaryForGeneration": parsed.get("summaryForGeneration") or "",
                "rawProviderResponse": raw,
            }
        )
        job_analysis_path.write_text(
            json.dumps(merged.model_dump(), indent=2) + "\n", encoding="utf-8"
        )
        return AnalyzeStepResult(job_analysis_path=job_analysis_path, via="perplexity")

    manual_prompt_path = inputs_dir / "job_analysis.MANUAL.md"
    manual_prompt = (
        "<!--\n"
        "Manual Step 1: paste this prompt into Perplexity (or any chat UI),\n"
        "then save the JSON response to inputs/job_analysis.json next to this file.\n"
        "The CLI's `retrieve` and `rank` subcommands will pick it up automatically.\n"
        "-->\n\n"
    ) + user_prompt
    manual_prompt_path.write_text(manual_prompt, encoding="utf-8")

    if not job_analysis_path.exists():
        stub = JobAnalysis(
            schemaVersion=1,
            generatedAt=now_iso(),
            jobSource=job_rel_path,
            tagsForGraphQL=TagsForGraphQL(
                domains=[],
                stack=[],
                kinds=["experience", "education", "project"],
                groups=[],
            ),
            tagsForTokenization=[],
            summaryForGeneration="TODO: fill summaryForGeneration via the manual prompt.",
        )
        job_analysis_path.write_text(
            json.dumps(stub.model_dump(), indent=2) + "\n", encoding="utf-8"
        )

    return AnalyzeStepResult(
        job_analysis_path=job_analysis_path,
        via="manual",
        manual_prompt_path=manual_prompt_path,
    )
